R/stackingutilities.R

```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+
+    if (!is.data.frame(text)) {
+        text <- as.data.frame(text, optional = TRUE)
+        text.names <- TidyLabels(colnames(text))
+        colnames(text) <- text.names
+        attr(text, "codeframe") <- fakeCodeFrame(text.names)
+        attr(text, "questiontype") <- "TextMulti"
+        question.names <- vapply(text, FUN = function(x) attr(x, "question"), FUN.VALUE = character(1))
+        text.label <- ExtractCommonPrefix(question.names)$common.prefix
+    } else {
+        text.label <- attr(text, "question")
+    }
+    
+    n.text.vars <- ncol(text)
+
+    if (is.null(existing.categorization)) { 
+        # No existing categorization to match against.
+        # Stack the text and the filter and weight
+        st <- stack(lapply(text, as.vector))
+        inds <- paste0(rownames(st), ".", st[, "ind"])
+        input.data <- list(text = st[, "values"],
+                          existing = NULL,
+                          subset = if (length(subset) == 1) subset else rep(subset, times = n.text.vars),
+                          weights = rep(weights, times = n.text.vars),
+                          inds = inds)
+    } else {
+        text <- prepareTextVariableLabelsForStackingWithGrids(text, existing.categorization)
+        if (attr(existing.categorization, "questiontype") == "PickAnyGrid") {
+            text.names <- names(attr(text, "codeframe"))
+            stacking <- ProcessAndStackDataForRegression(list(Y = text, X = existing.categorization), 
+                                                            formula = NULL, 
+                                                            interaction = NULL, 
+                                                            subset = subset, 
+                                                            weights = weights)
+            colnames(stacking$data) <- vapply(stacking$data, 
+                                            FUN = function (x) attr(x, "label"), 
+                                            FUN.VALUE = character(1))
+            text <- stacking$data[, 1]
+            existing <- stacking$data[, 2:length(stacking$data)]
+            colnames(existing) <- TidyLabels(colnames(existing))
+            input.data <- list(text = text,
+                                existing = existing,
+                                subset = stacking$subset,
+                                weights = stacking$weights,
+                                inds = rownames(stacking$data))
+        } else if (attr(existing.categorization, "questiontype") == "PickOneMulti") {
+            # Reorder the variables in the categorization to match those
+            # of the text.
+            m <- match(colnames(existing.categorization), colnames(text))
+            existing.categorization <- existing.categorization[, m]
+            st <- stack(lapply(text, as.vector))
+            inds <- paste0(rownames(st), ".", st[, "ind"])
```

Comment:
similar update to above, change `inds` so that it is `case_number.variable` instead of `1:N.variable` where `N = n * p`.